File: arrays/DIStringMatch.js
/**
 * @param {string} S
 * @return {number[]}
 */
var diStringMatch = function(S) {
    var large = S.length;
    var small = 0;
    var ans = [];
    
    S.split("").forEach((s, i) => {
        if (s === "D") {
            ans.push(large);
            large--;
        } else {
            ans.push(small);
            small++;
        }
        // if it is the last one
        if (i === S.length - 1) {
            if (s === 'D') {
                ans.push(small);       
            } else {
                ans.push(large);
            }
        }
    });
    
    return ans;
};

File: arrays/DIStringMatch.go
func diStringMatch(S string) []int {
    var small int = 0
    var large int = len(S)
    var ans = make([]int, len(S) + 1)
    
    for i,e := range S {
        if string(e) == "I" {
            ans[i] = small
            small++
        } else {
            ans[i] = large
            large--
        }
        
        if (i == len(S) - 1) {
            if string(e) == "I" {
                ans[i+1] = small
            } else {
                ans[i+1] = large
            }                
        }
    }
    return ans
}